runtime/src/message_processor.rs

```diff
@@ -296,6 +303,28 @@ impl<'a> ThisInvokeContext<'a> {
             instruction_recorder,
             feature_set,
             timings: ExecuteDetailsTimings::default(),
+            account_db,
+            ancestors,
+            sysvars: vec![],
+        }
+    }
+    fn load_sysvar(&self, pubkey: &Pubkey) -> Option<Rc<Sysvars>> {
+        let account = self
+            .account_db
+            .load_slow(self.ancestors, pubkey)
+            .map(|(account, _)| Rc::new(RefCell::new(account)))?;
+        let borrowed = &account.try_borrow_mut().ok()?;
+
+        if &clock::id() == pubkey {
+            Some(Rc::new(Sysvars::Clock(
+                bincode::deserialize(&borrowed.data()[..]).ok()?,
+            )))
+        } else if &rent::id() == pubkey {
+            Some(Rc::new(Sysvars::Rent(
+                bincode::deserialize(&borrowed.data()[..]).ok()?,
+            )))
```

Comment:
I think the `bincode::deserialize()` here is unnecessary.   The data ends up getting unsafely cast back to a buffer for transport across the syscall boundary anyway, so it looks like the `Sysvars` struct could be removed entirely leaving us with `sysvars: Vec<(Pubkey, Option<Rc<Vec<u8>>>)>,`.

Reply:
I originally was just generically passing the data back across the syscall boundary as `u8` but wanted to avoid doing the `bincode::deserialize()` in BPF program space.  Was much simpler codewise passing u8 but performance-wise not so much.  Also, the current approach allows us to cache the deserialized value.

Reply:
> I originally was just generically passing the data back across the syscall boundary as u8 but wanted to avoid doing the bincode::deserialize() in BPF program space. 

The program needs to `bincode::deserialize()` if the sysvar is passed in as an account though right, so we're not making things worse.  Keeping the interface consistent  between the now two paths to obtain a sysvar seems desirable and, better, should be much simpler code wise

>  Also, the current approach allows us to cache the deserialized value.

But the program really shouldn't be calling the syscall multiple times (except for perhaps some library code that a program imported)

Reply:
The problem here is that `bincode::deserialize()` doesn't work for the large sysvars and `StakeHistory` is slippy by right now because the history isn't very long.  Pretty soon it is going to grow too large to deserrialize in program.

I see two options:
- map the sysvar into the program space and provide helpers that interact with it (`get_num_stakes` and `get_stake_at_index` come to mind).
- export syscalls on a per-syscall basis that does the same as the helpers I just mentioned